core/src/main/java/com/sequenceiq/cloudbreak/service/stack/flow/FinalizationCleanUpService.java

```diff
@@ -0,0 +1,77 @@
+import com.sequenceiq.cloudbreak.domain.stack.cluster.Cluster;
+import com.sequenceiq.cloudbreak.domain.stack.cluster.ClusterComponentHistory;
+import com.sequenceiq.cloudbreak.service.cluster.ClusterService;
+import com.sequenceiq.cloudbreak.service.stack.StackService;
+import com.sequenceiq.cloudbreak.structuredevent.db.LegacyStructuredEventDBService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Service;
+
+@Service
+public class FinalizationCleanUpService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(FinalizationCleanUpService.class);
+
+    private final StackService stackService;
+
+    private final ClusterService clusterService;
+
+    private final LegacyStructuredEventDBService structuredEventDBService;
+
+    private final ClusterComponentConfigProvider clusterComponentConfigProvider;
+
+    public FinalizationCleanUpService(StackService stackService, ClusterService clusterService,
+            LegacyStructuredEventDBService structuredEventDBService,
+            ClusterComponentConfigProvider clusterComponentConfigProvider) {
+        this.stackService = stackService;
+        this.clusterService = clusterService;
+        this.structuredEventDBService = structuredEventDBService;
+        this.clusterComponentConfigProvider = clusterComponentConfigProvider;
+    }
+
+    public void detachClusterComponentRelatedAuditEntries() throws CleanUpException {
+        LOGGER.debug("About to start fetch and clean up detached/orphaned cluster component audit events.");
+        boolean success = true;
+        try {
+            clusterComponentConfigProvider.cleanUpDetachedEntries();
+        } catch (Exception e) {
+            LOGGER.info("Unable to clean up audit entries due to: " + e.getMessage(), e);
+            success = false;
+            throw new CleanUpException(ClusterComponentHistory.class.getSimpleName() + " cleanup has failed!", e);
+        } finally {
+            String postfix = success ? "successfully." : "with error.";
+            LOGGER.debug("Fetching and cleaning up detached/orphaned audit events has finished {}", postfix);
+        }
+    }
+
+    public void cleanUpStructuredEvents() {
+        LOGGER.debug("About to start fetch and clean up structured events.");
+        boolean success = true;
+        try {
+            Set<Long> idsForChecking = new HashSet<>();
+            idsForChecking.addAll(clusterService.findDeletedClusters().stream().map(Cluster::getId).collect(toSet()));
+            idsForChecking.addAll(stackService.findDeletedStackIds());
+            LOGGER.debug("About to request the deletion of {} {} that belogs to the following resource IDs: {}",
+                    StructuredEventEntity.class.getSimpleName(), idsForChecking.size() > 1 ? "entries" : "entry",
+                    String.join(",", idsForChecking.stream().map(Object::toString).collect(toSet())));
```

Comment:
I think (but you an experiment as well) that if you have a Set type variable and you call toString() on it, that will produce the same (or similar) result, good enough for logging.